crates/lair_keystore_api/src/lair_store.rs

```diff
@@ -0,0 +1,420 @@
+//! lair persistance
+
+use crate::prelude::*;
+use futures::future::BoxFuture;
+use std::future::Future;
+use std::sync::Arc;
+
+/// Helper traits for store types - you probably don't need these unless
+/// you are implementing new lair core instance logic.
+pub mod traits {
+    use super::*;
+
+    /// Defines a lair storage mechanism.
+    pub trait AsLairStore: 'static + Send + Sync {
+        /// Return the context key for both encryption and decryption
+        /// of secret data within the store that is NOT deep_locked.
+        fn get_bidi_ctx_key(&self) -> sodoken::BufReadSized<32>;
+
+        /// List the entries tracked by the lair store.
+        fn list_entries(
+            &self,
+        ) -> BoxFuture<'static, LairResult<Vec<LairEntryInfo>>>;
+
+        /// Write a new entry to the lair store.
+        fn write_entry(
+            &self,
+            entry: LairEntry,
+        ) -> BoxFuture<'static, LairResult<()>>;
```

Comment:
note the single generic write_entry fn for implementing a store backend, where the concrete wrapper below includes logic for writing specific entry types (seed, deep_locked_seed, tls_cert, etc).